xblock/field_data.py

```
@@ -161,7 +161,8 @@ def set_many(self, block, update_dict):
         for key, value in update_dict.items():
             update_dicts[self._field_data(block, key)][key] = value
         for field_data, new_update_dict in update_dicts.items():
-            field_data.set_many(block, new_update_dict)
+            if field_data is not None:  # Ignore fields from scopes that are not loaded
```

Do you think we should log a warning in this case?

---

Yeah, I don't think this error should be silently ignored at this level. In fact, I think raising an exception is the right behavior. If we choose to ignore unsupported scopes like `Scope.user_state`, we should be doing that explicitly by mapping that scope to a no-op.

> As the `LearningCoreXBlockRuntime` doesn't have the `Scope.user_state`, the get_block function throws an error.

I don't really understand this. [We _do_ support the `Scope.user_state` scope in Learning Core runtime](https://github.com/openedx/edx-platform/blob/c84af052517b46c95a6c0600fe9bbace920188f0/openedx/core/djangoapps/xblock/runtime/runtime.py#L385). Although I just noticed now that there is one circumstance where we don't: when `user is None`:

https://github.com/openedx/edx-platform/blob/c84af052517b46c95a6c0600fe9bbace920188f0/openedx/core/djangoapps/xblock/runtime/runtime.py#L353-L367

So: I suspect the root problem here is that **you're trying to make changes to `choices` when `user` is `None`**, which makes no sense. But if `user` is set, then something else weird is going on. In any case, I don't think ignoring it is the right fix.

---

Ah, I see now we're not really modifying any fields, just marking them as loaded.

One fix could be to change [this line](https://github.com/openedx/edx-platform/blob/c84af052517b46c95a6c0600fe9bbace920188f0/openedx/core/djangoapps/xblock/runtime/learning_core_runtime.py#L225) to omit any fields from user-specific scopes if `user=None`, but that seems fragile.

A simpler approach is probably to change [this line](https://github.com/openedx/edx-platform/blob/c84af052517b46c95a6c0600fe9bbace920188f0/openedx/core/djangoapps/xblock/runtime/runtime.py#L355) so that instead of setting `student_data_store = None` , we set `student_data_store` to [`DictFieldData`](https://github.com/openedx/XBlock/blob/03251b23500183f2b6ee32e056c0447cf812753a/xblock/field_data.py#L104) to explicitly ignore any writes there. It could be a subclass that logs warnings, but the true root cause here is that XBlock Field Data API requires calling `set` when initializing the block, so we'd be issuing a pointless warning because nobody's actually trying to write to that field; the field simply exists and we're loading the initial value.

---

The underlying issue is that the `choices` field is marked as `dirty` on load. I'll dig a bit deeper and try to fix that.

In retrospect, I think we should throw an exception if the scope is not set on `scope_mappings`. We have other blocks that throw the inconclusive `AttributeError: 'NoneType' object has no attribute 'set_many'` (i.e. `problem-builder` because of `Scope.children` being `None`).

---

Yeah, XBlock fields are marked dirty on read if their types are mutable, i.e. lists, sets, dicts. I have a [rant blog post](https://medium.com/@dormsbee/xblock-lessons-costly-conveniences-38ff8e7de6f1) that talks about this.

---

Thanks @ormsbee!
I just found that. :disappointed:

Not sure the best way to go from here:
1. Clean `user_state_*` fields from `get_fields_to_save()` on load
2. Use the 'DictFieldData` as `student_data_store`
3. ????

---

Hi @bradenmacdonald! Do you think we should go on the `DictFieldData` route?

---

@rpenido I'm leaning that way. I think it's a relatively simple fix for the problem you're seeing.

---

Thanks @bradenmacdonald! PR created: https://github.com/openedx/edx-platform/pull/36226
I changed this current PR to throw `InvalidScopeError` instead of `AttributeError` if the scope is not set (it will not happen anymore for the block load, but it can occur in other cases).